core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessageDuplicator.java

```diff
@@ -388,8 +331,8 @@ private void doUnsubscribe(DownstreamSubscription<T> subscription, @Nullable Thr
         }
 
         private void doCleanupIfLastSubscription() {
-            if (isLastDownstreamAdded() && downstreamSubscriptions.isEmpty()) {
-                state = State.CLOSED;
+            if (isClosed() && duplicator.unsubscribed == 0 && downstreamSubscriptions.isEmpty()) {
```

Comment:
The thread safety of this `unsubscribed == 0` check isn't so obvious so I think we can add a comment. Maybe

`// Because the duplicator is closed, we know that unsubscribed will not be incremented anymore and are guaranteed that the last unsubscribed downstream will run this cleanup logic.`

Reply:
That's a good suggestion. Added. 😄

Reply:
Anyway, I am working on #2189 and it will be done today or tomorrow. (I hope 😆 )